### ISSUE TITLE
Binding to integer does not differentiate positive from negative values

### ISSUE THREAD

**Issue by VicenteRD** [spring-projects/spring-boot#14136]
Hello, I've recently migrated to Boot 2.0 and my working app now has a very weird issue.

I have a `Map<String, Map<Integer, AttributeLink>> links` that gets loaded from a YML file, where `AttributeLink` contains a `UserRank rank` (enum loaded from int), `Integer group` and `List<Long> roleIds`.
All in all, my config looks like this:
```
    patreon:                   <-    Attribute
      1:                       <-    Tier
        rank:                  ---|
        group:                    |- AttributeLink
        roleIds:                  |
          - 112884142022078464 ---|
```

This used to work just fine, until I migrated. When I migrated, the `roleId` list under tier `-1` would be the copy of the list under tier `1`. If I changed the tier to `-2`, it would take the value of the list on tier `2`, basically mirroring the positive tier list. The other 2 attributes (`rank` and `group`) would not be mirrored. My solution was to turn `-1` into `99` and change the logic a bit.
Let me know if anything else is needed.

**Comment by VicenteRD** [comment]
Upon further testing, it can happen both ways (- on +, + on -), but I am not sure on the conditions for it to happen. It'd seem the only thing unaffected is the `UserRank` enum (testing this is a bit more difficult).

Sadly, it's a private repo. But I can share this simplified version of [the class](https://hastebin.com/habuyijaxo.java) that still has the issue with a [test config](https://hastebin.com/yililazita.yml) which I use alongside a `application.properties`.

**Comment by snicoll** [comment]
It doesn't have to be your real project. All we need is a simple project that showcases what you mean (anything that's not related to the problem is irrelevant and doesn't need to be part of the sample anyway). Thanks again.

**Comment by VicenteRD** [comment]
I'll set something up within the day then, thanks for the patience!

**Comment by snicoll** [comment]
Thanks for the sample.

It looks like that integer binding does not differentiate positive and negative values. That's probably due to the relaxed binding changes that massages the input key. I've tried to wrap the keys in `[]` and changing to Strings but that didn't help. Switching that project to `1.5.14.RELEASE` works so I confirm the regression (ignoring the very unusual binding for enum, I wasn't even aware we supported that!)
